src/zeroband/train.py

```diff
@@ -176,13 +176,139 @@ def get_logprobs(model: ModelType, input_ids: torch.Tensor, position_ids: torch.
     return logprobs
 
 
+def log_to_wandb(metrics_dict, tokenizer=None, batch=None, step=None):
+    """Log metrics to wandb with proper grouping."""
+    # Initialize grouped metrics dictionary
+    wandb_metrics = {}
+
+    # Group metrics by category
+    categories = {
+        "train/": ["step", "rollout_step", "inner_lr", "total_tokens", "total_problems"],
+        "losses/": ["Loss", "pg_loss", "entropy_loss", "kl", "grad_norm", "clip_ratio"],
+        "rewards/": ["sample_reward", "task_reward", "batch_reward", "batch_task_reward"],
+        "lengths/": [
+            "seq_lens",
+            "batch_seq_lens",
+            "target_lengths",
+            "batch_target_lengths",
+            "padding_proportion",
+            "length_penalties",
+            "batch_length_penalties",
+        ],
+        "perf/": [
+            "tokens_per_second",
+            "tokens_per_second_per_gpu",
+            "mfu",
+            "time_rollout_step",
+            "time_logprob",
+            "time_data_loading",
+            "time_packing",
+        ],
+    }
+
+    # Add metrics to respective groups
+    for prefix, keys in categories.items():
+        for k in keys:
+            if k in metrics_dict and metrics_dict[k] is not None:
+                wandb_metrics[f"{prefix}{k}"] = metrics_dict[k]
+
+    # Log everything
+    wandb.log(wandb_metrics, step=metrics_dict.get("step", step))
+
+
+def log_prompt_response_samples(tokenizer, batch, step, sample_history=None):
+    """Log samples using wandb.Table with accumulated history.
+    Only logs every 5 steps to reduce overhead.
```

Comment:
same lets move  this to src/zeroband/training/utils.py

Reply:
also done